libutils/threaded_deque.c

```diff
@@ -0,0 +1,641 @@
+    int ret = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to use error-checking mutexes for deque, "
+            "falling back to normal ones (pthread_mutexattr_settype: %s)",
+            GetErrorStrFromCode(ret));
+        pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_NORMAL);
+    }
+
+    new_deque->lock = xmalloc(sizeof(pthread_mutex_t));
+    ret = pthread_mutex_init(new_deque->lock, &attr);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize mutex (pthread_mutex_init: %s)",
+            GetErrorStrFromCode(ret));
+        pthread_mutexattr_destroy(&attr);
+        free(new_deque->lock);
+        free(new_deque);
+        return NULL;
+    }
+
+    new_deque->cond_non_empty = xmalloc(sizeof(pthread_cond_t));
+    ret = pthread_cond_init(new_deque->cond_non_empty, NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize thread condition "
+            "(pthread_cond_init: %s)",
+            GetErrorStrFromCode(ret));
+        free(new_deque->lock);
+        free(new_deque->cond_non_empty);
+        free(new_deque);
+        return NULL;
+    }
+
+    new_deque->cond_empty = xmalloc(sizeof(pthread_cond_t));
+    ret = pthread_cond_init(new_deque->cond_empty, NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize thread condition "
+            "(pthread_cond_init: %s)",
+            GetErrorStrFromCode(ret));
+        free(new_deque->lock);
+        free(new_deque->cond_empty);
+        free(new_deque->cond_non_empty);
+        free(new_deque);
+        return NULL;
+    }
+
+    return new_deque;
+}
+
+/**
+  @brief Destroys data in range.
+  @warning Assumes that locks are acquired.
+  @note If start == end, this means that all elements in deque will be
+        destroyed. Since the internal array is circular, it will wrap around
+        when reaching the array bounds.
+  @param [in] deque  Pointer to struct.
+  @param [in] start  Position to start destroying from.
+  @param [in] end    First position to not destroy. Can be same as start to
+                     destroy the entire allocated area.
+  */
+static void DestroyRange(ThreadedDeque *deque, size_t start, size_t end)
+{
+    assert(deque != NULL);
+    if (start > deque->capacity || end > deque->capacity)
+    {
+        Log(LOG_LEVEL_DEBUG,
+            "Failed to destroy ThreadedDeque, index greater than capacity: "
+            "start = %zu, end = %zu, capacity = %zu",
+            start, end, deque->capacity);
+        return;
+    }
+
+    if ((deque->ItemDestroy != NULL) && deque->size > 0)
+    {
+        size_t i = start;
+
+        do
+        {
+            deque->ItemDestroy(deque->data[i++]);
+            i %= deque->capacity;
+        } while (i != end);
+    }
+}
+
+/**
+  @brief Either expands capacity of dedeque, or shifts tail to beginning.
```

Comment:
*dedeque*

Reply:
Heh, some of this is directly from `s/queue/deque/g`.